Implement aoc:u and support loading AddOnContent

mageven opened the pull request:
Implements most of `IAddOnContentManager` based on fw 9.2.0 RE and a little bit from 10.0.0.
~~Also adds support for reading AddOnContent embedded in XCI.~~

This is my first RE work in this project so please be extra critical.

Supersedes #1218 (Sorry @AcK77 !)
Fixes #636

**EDIT:**
Now supports reading embedded AOC from XCI and NSP. Also, automatically loads any AOC NSPs in the game's base directory.
With some GUI work, custom AOC paths can be added and toggled.

~~Depends on #1236~~

gdkchan commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,190 @@
+        {
+            ulong pid = context.RequestData.ReadUInt64();
+
+            // Official code calls arp:r GetApplicationControlProperty to get AddOnContentBaseId
```
Libhac does have a ARP implementation now, I wonder if we could just call it.

Thealexbarney replied:
Libhac has support for `arp:r` functions in its nnsdk-equivalent client class. The service implementation still needs to be provided which is what [this](https://github.com/Ryujinx/Ryujinx/blob/4b79b78b8c3b9cd6deed46e6aa38acabfc47d25b/Ryujinx.HLE/HOS/Services/Arp/LibHacIReader.cs) does. Note: it's a very simple implementation since Ryujinx currently only keeps track of the current application and doesn't have a concept of NCM storages

gdkchan replied:
Yea I noticed that later, and `GetApplicationControlProperty` is not implemented at all, so nevermind my comment.

gdkchan commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,190 @@
+        // [10.0.0+] GetAddOnContentLostErrorCode() -> u64
+        public static ResultCode GetAddOnContentLostErrorCode(ServiceCtx context)
+        {
+            // Seems to read from static addr=&(0x7d0a4 << 32)?
```
This address doesn't make much sense 🤔

mageven replied:
Yeah, messed up on a pointer dereference. I corrected it now.

mageven commented:
Now supports reading embedded AOC from XCI and NSP. Also, automatically loads any AOC NSPs in the game's base directory.

@gdkchan Sorry had to rebase it for sanity. AOC code starts from 3cff410.

Maybe needs a `blocked` label

mageven commented:
Rebased to master.

mageven commented:
Reverted some renaming. This should allow merging with Mods PR without conflicts.
Also rebased to latest master as some crucial fixes went in.

mageven commented:
Yes, the current auto-scan is temporary (also serves as an example to show usage). But, I hope even in the eventual GUI, a more general scanning is added is some way (like in `LoadApplications`) and the GUI just shows toggles on scanned addons, instead of just add/remove per title.

AcK77 commented on Ryujinx.HLE/FileSystem/Content/ContentManager.cs:
```diff
@@ -176,6 +196,80 @@ void AddEntry(LocationEntry entry)
+                    }
+
+                    using var pfs0 = nca.OpenFileSystem(0, Switch.GetIntegrityCheckLevel());
+                    pfs0.OpenFile(out IFile cnmtFile, pfs0.EnumerateEntries().Single().FullPath.ToU8Span(), OpenMode.Read);
```
Maybe add a new empty line before and after to make things a bit more readable?

AcK77 commented on Ryujinx.HLE/FileSystem/Content/ContentManager.cs:
```diff
@@ -176,6 +196,80 @@ void AddEntry(LocationEntry entry)
+                    if (nca.Header.ContentType != NcaContentType.Meta)
+                    {
+                        Logger.PrintWarning(LogClass.Application, $"{ncaPath} is not a valid metadata file");
+                        continue;
```
Add a new empty line before

AcK77 commented on Ryujinx.HLE/FileSystem/Content/ContentManager.cs:
```diff
@@ -176,6 +196,80 @@ void AddEntry(LocationEntry entry)
+
+        public bool GetAocDataStorage(ulong aocTitleId, out IStorage aocStorage)
+        {
+            aocStorage = null;
```
Add a new empty line after

AcK77 commented on Ryujinx.HLE/FileSystem/Content/ContentManager.cs:
```diff
@@ -176,6 +196,80 @@ void AddEntry(LocationEntry entry)
+                var file = new FileStream(aoc.ContainerPath, FileMode.Open, FileAccess.Read);
+                PartitionFileSystem pfs;
+                IFile ncaFile;
+                switch (Path.GetExtension(aoc.ContainerPath))
```
Add a new empty line before

AcK77 commented on Ryujinx.HLE/FileSystem/Content/ContentManager.cs:
```diff
@@ -176,6 +196,80 @@ void AddEntry(LocationEntry entry)
+                    default:
+                        return false; // Print error?
+                }
+                aocStorage = new Nca(_virtualFileSystem.KeySet, ncaFile.AsStorage()).OpenStorage(NcaSectionType.Data, Switch.GetIntegrityCheckLevel());
```
Add a new empty line before and after

AcK77 commented on Ryujinx.HLE/HOS/ApplicationLoader.cs:
```diff
@@ -144,6 +144,14 @@ public void LoadXci(string xciFile)
+            _contentManager.AddAocData(securePartition, xciFile, mainNca.Header.TitleId);
+
+            // Check all nsp's in the base directory for AOC
+            foreach(var fn in new FileInfo(xciFile).Directory.EnumerateFiles("*.nsp"))
```
```suggestion
            foreach (var fn in new FileInfo(xciFile).Directory.EnumerateFiles("*.nsp"))
```

AcK77 commented on Ryujinx.HLE/HOS/ApplicationLoader.cs:
```diff
@@ -177,6 +185,15 @@ public void LoadNsp(string nspFile)
+                // Check all nsp's in the base directory for AOC
+                foreach(var fn in new FileInfo(nspFile).Directory.EnumerateFiles("*.nsp"))
+                {
+                    if(fn.FullName == nspFile) continue;
```
Add a new empty line after

AcK77 commented on Ryujinx.HLE/HOS/ApplicationLoader.cs:
```diff
@@ -177,6 +185,15 @@ public void LoadNsp(string nspFile)
+                _contentManager.AddAocData(nsp, nspFile, mainNca.Header.TitleId);
+
+                // Check all nsp's in the base directory for AOC
+                foreach(var fn in new FileInfo(nspFile).Directory.EnumerateFiles("*.nsp"))
```
```suggestion
                foreach (var fn in new FileInfo(nspFile).Directory.EnumerateFiles("*.nsp"))
```

AcK77 commented on Ryujinx.HLE/HOS/Services/Fs/IFileSystemProxy.cs:
```diff
@@ -372,17 +372,24 @@ public ResultCode OpenDataStorageByDataId(ServiceCtx context)
+            if (context.Device.System.ContentManager.GetAocDataStorage((ulong)titleId, out LibHac.Fs.IStorage aocStorage))
+            {
+                Logger.PrintInfo(LogClass.Loader, $"Opened AddOnContent Data TitleID={titleId:X16}");
+                MakeObject(context, new FileSystemProxy.IStorage(aocStorage));
```
Add a new empty line before and after

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
         public static ResultCode CountAddOnContent(ServiceCtx context)
         {
-            context.ResponseData.Write(0);
+            ulong pid = context.RequestData.ReadUInt64();
```
Since we don’t support multiprocess for now, we usually don’t read the pid through the request data, you can get the current Pid directly in a context field.

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
+
+            context.ResponseData.Write(aocCount);
+
+            Logger.PrintDebug(LogClass.ServiceNs, $"pid={pid} count={aocCount} RuntimeInstall={runtimeAddOnContentInstall}");
```
Our PrintDebug() don’t support include the var directly and print the names using reflection?

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
 namespace Ryujinx.HLE.HOS.Services.Ns
 {
     [Service("aoc:u")]
     class IAddOnContentManager : IpcService
```
All calls don’t have to be static if you don’t call them outside of the class, in case you have to call some method we usually use a static submethod!

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
+            if (aocCount <= startIndex)
+            {
+                context.ResponseData.Write((uint)0);
+                return ResultCode.Success;
```
Add a new empty line before a return!

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
+                context.Memory.Write(bufAddr + (ulong)i * 4, (int)(aocTitleIds[i + (int)startIndex] - aocBaseId));
+            }
+
+            Logger.PrintDebug(LogClass.ServiceNs, $"pid={pid} bufferSize={bufferSize} start={startIndex} aocCount={aocCount}");
```
Same as above

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
+        // GetAddOnContentBaseId(pid) -> u64
+        public static ResultCode GetAddonContentBaseId(ServiceCtx context)
+        {
+            ulong pid = context.RequestData.ReadUInt64();
```
Same as above

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
+        public static ResultCode PrepareAddOnContent(ServiceCtx context)
+        {
+            uint aocIndex = context.RequestData.ReadUInt32();
+            ulong pid = context.RequestData.ReadUInt64();
```
Same as above

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
+        }
+
+
+        [Command(9)]
```
```suggestion
        [Command(9)] // 10.0.0+
```

AcK77 commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,188 @@
+
+
+        [Command(9)]
+        // [10.0.0+] GetAddOnContentLostErrorCode() -> u64
```
```suggestion
        // GetAddOnContentLostErrorCode() -> u64
```

mageven commented:
Addressed and rebased. Review replies here.
> Since we don’t support multiprocess for now, we usually don’t read the pid through the request data, you can get the current Pid directly in a context field.

I hope the new way is fine. Still kept the local vars, just to show that the function captures it.

> Our PrintDebug() don’t support include the var directly and print the names using reflection?

Sadly, only `PrintStub` has overloads for object reflections. I now realize that `PrintDebug` isn't much useful, so used `PrintStub` as it's in-line with other services.

> All calls don’t have to be static if you don’t call them outside of the class, in case you have to call some method we usually use a static submethod!

Hm. Don't remember why I did this. Maybe copied all of them from one method.

Thog commented on Ryujinx.HLE/HOS/ApplicationLoader.cs:
```diff
@@ -177,6 +185,16 @@ public void LoadNsp(string nspFile)
+                {
+                    if (fn.FullName == nspFile) continue;
+                    
+                    _contentManager.AddAocData(new PartitionFileSystem(fn.OpenRead().AsStorage()), fn.FullName, mainNca.Header.TitleId);
```
Shouldn't the PartitionFileSystem **must** be disposed at some point.

Thog commented on Ryujinx.HLE/HOS/ApplicationLoader.cs:
```diff
@@ -144,6 +144,14 @@ public void LoadXci(string xciFile)
+            // Check all nsp's in the base directory for AOC
+            foreach (var fn in new FileInfo(xciFile).Directory.EnumerateFiles("*.nsp"))
+            {
+                _contentManager.AddAocData(new PartitionFileSystem(fn.OpenRead().AsStorage()), fn.FullName, mainNca.Header.TitleId);
```
Shouldn't the PartitionFileSystem **must** be disposed at some point.

Thog commented on Ryujinx.HLE/HOS/Services/Ns/IAddOnContentManager.cs:
```diff
@@ -1,34 +1,189 @@
 
             return ResultCode.Success;
         }
+
```
nit extra line return

mageven commented:
Addressed, rebased and fixed a potential bug. Tested with and without PTC. All seems fine.
